### PR TITLE
executor, types: copy enum/set instead of refer the underlying data

### LINKED ISSUE #16854: group by on enum column got wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
``` sql
create table t(a enum('WAITING','PRINTED','STOCKUP','CHECKED','OUTSTOCK','PICKEDUP','WILLBACK','BACKED'));

insert into t values(1),(2),(3),(4),(5),(6),(7);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;

set @@tidb_max_chunk_size=100;
```
### 2. What did you expect to see?
stable result

### 3. What did you see instead?
wrong result
``` sql
tidb> select distinct a from t;
+----------+
| a        |
+----------+
| RINTED  |
| TSTOCK  |
| WAITING |
| ECKED   |
| WILLBACK |
| OCKUP   |
| ICKEDUP  |
+----------+
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
``` sql
tidb> select tidb_version();
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                              |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-343-ge607157ef-dirty
Git Commit Hash: e607157efef0f76878d18a667b233061ce43545a
Git Branch: master
UTC Build Time: 2020-04-27 03:32:22
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #16854  

Problem Summary:

### What is changed and how it works?

What's Changed:
Copy enum/set instead of referring the underlying data in GetMysqlEnum and GetMysqlSet

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
    - Consumes more MEM

### Release note 
Fix unstable result when select distinct enum column or set column.

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

File: types/datum.go [hunk 1]
```diff
@@ -294,7 +294,7 @@ func (d *Datum) SetMysqlEnum(b Enum, collation string) {
 
 // GetMysqlSet gets Set value
 func (d *Datum) GetMysqlSet() Set {
-	str := string(hack.String(d.b))
+	str := string(d.b)
```
Comment (SunRunAway) [inline types/datum.go:297]: Datum.GetBytes returns reference either.
Better to modify the code in aggregation parts.

**Pull request merged**